### PR TITLE
[BEAM-837] Add authentication support in JmsIO

### PR DESCRIPTION
**Pull request by jbonofre**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
Rebased and changed tests for the authentication.

**Comment by jbonofre** [comment]
R: @dhalperi

**Comment by jbonofre** [comment]
Rebased to find the build.

CC: @davorbonaci

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java [hunk 1]
```diff
@@ -211,6 +215,23 @@ public Read withTopic(String topic) {
...
+    public Read withUsername(String username) {
+      checkArgument(username != null, "JmsIO.read().withUsername(username) called with null "
+          + "username");
+      return builder().setUsername(username).build();
+    }
+
+    /**
+     * Define the password to connect to the JMS broker (authenticated).
+     */
+    public Read withPassword(String password) {
```
Comment (dhalperi) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: Make it `@Nullable` if it is allowed to be `null`?

Though I would maybe think that if someone runs this function, it should not be allowed to be `null`. They're deliberately trying to supply a password, so we should catch it if they send `null` instead.

(basically, up to you: do you want users to be able to "clear" the password here using `null`?)

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java [hunk 2]
```diff
@@ -369,17 +390,23 @@ public UnboundedJmsReader(
...
     @Override
     public boolean start() throws IOException {
-      ConnectionFactory connectionFactory = source.spec.getConnectionFactory();
+      Read spec = source.spec;
+      ConnectionFactory connectionFactory = spec.getConnectionFactory();
       try {
-        this.connection = connectionFactory.createConnection();
+        if (spec.getUsername() != null) {
+          this.connection =
+              connectionFactory.createConnection(spec.getUsername(), spec.getPassword());
```
Comment (dhalperi) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: Not clear to me from the [Javadoc](https://docs.oracle.com/javaee/7/api/javax/jms/ConnectionFactory.html#createConnection-java.lang.String-java.lang.String-) whether password is allowed to be `null`. Can you check? If so, you might want to try `MoreObjects.firstNonNull(password, "");`

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java [hunk 3]
```diff
@@ -572,6 +603,23 @@ public Write withTopic(String topic) {
...
+      checkArgument(username != null,  "JmsIO.write().withUsername(username) called with null "
+          + "username");
+      return builder().setUsername(username).build();
+    }
+
+    /**
+     * Define the password to connect to the JMS broker (authenticated).
+     */
+    public Write withPassword(String password) {
+      // password can be null
```
Comment (dhalperi) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: same comment here.

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java [hunk 4]
```diff
@@ -601,7 +649,13 @@ public WriterFn(Write spec) {
       @StartBundle
       public void startBundle(Context c) throws Exception {
         if (producer == null) {
-          this.connection = spec.getConnectionFactory().createConnection();
+          if (spec.getUsername() != null) {
+            this.connection =
+                spec.getConnectionFactory()
+                    .createConnection(spec.getUsername(), spec.getPassword());
+          } else {
+            this.connection = spec.getConnectionFactory().createConnection();
```
Comment (dhalperi) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: and here.

File: sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java [hunk 5]
```diff
@@ -58,13 +63,26 @@
...
     broker.setUseJmx(false);
     broker.setPersistenceAdapter(new MemoryPersistenceAdapter());
     broker.addConnector(BROKER_URL);
     broker.setBrokerName("localhost");
+    broker.setPopulateJMSXUserID(true);
+    broker.setUseAuthenticatedPrincipalForJMSXUserID(true);
+
+    // enable authentication
+    List<AuthenticationUser> users = new ArrayList<>();
+    users.add(new AuthenticationUser("test", "test", "users,admins"));
```
Comment (dhalperi) [inline sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java]: can you document what parameters these are? e.g.,

`/* user */ test, /* password */ test, /* groups */ users, admins`

of course with the right parameter names and quotes, etc.

Comment (dhalperi) [inline sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java]: Can you make the username and password different, to ensure we put the right parameters in the right place?

Comment (dhalperi) [inline sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java]: Also, can you extract these into `private static final String TEST_USERNAME = test` and etc? and reuse those throughout the tests? Easier to change, lower chance of typos, etc. ;)

File: sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java [hunk 6]
```diff
@@ -78,10 +96,24 @@ public void stopBroker() throws Exception {
...
+    expectedException.expect(Exception.class);
+    expectedException.expectMessage("JMSSecurityException");
+
+    pipeline.apply(
+        JmsIO.read()
+            .withConnectionFactory(connectionFactory)
+            .withQueue("test"));
+
+    pipeline.run();
+  }
```
Comment (dhalperi) [inline sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java]: (up to you: maybe also a test with the right username but wrong password?)

File: sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java [hunk 7]
```diff
@@ -78,10 +96,24 @@ public void stopBroker() throws Exception {
 
   @Test
   @Category(NeedsRunner.class)
+  public void testAuthenticationRequired() {
+    expectedException.expect(Exception.class);
+    expectedException.expectMessage("JMSSecurityException");
```
Comment (dhalperi) [inline sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java]: Optional:

how good is the error message returned here? Does it say "you must log in"? or just it just does "JMSSecurityException"?

For user experience, we should try to make sure it's a good error. And if so, maybe we should test some key phrase like "log in" if there is a good one.

**Comment by jbonofre** [comment]
Rebased and updated according to Dan's comments.

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java [hunk 8]
```diff
@@ -616,7 +617,8 @@ public Write withUsername(String username) {
      * Define the password to connect to the JMS broker (authenticated).
      */
     public Write withPassword(String password) {
-      // password can be null
+      checkArgument(password != null, "JmsIO.write().withPasswprd(password) called with null "
```
Comment (dhalperi) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: typo: Password

**Comment by jbonofre** [comment]
@dhalperi Thanks ! Waiting the end of the checks before merging.

**Comment by jbonofre** [comment]
Travis seems to be "stuck". As Jenkins is OK, I gonna merge.

**Pull request closed without merging**